crates/bevy_ecs/src/query/access.rs

```diff
@@ -85,30 +47,265 @@ impl<T: SparseSetIndex> Default for Access<T> {
     }
 }
 
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) struct SortedSmallVec<const N: usize>(SmallVec<[usize; N]>);
+
+
+impl<const N: usize> IntoIterator for SortedSmallVec<N> {
+
+    type Item = usize;
+    type IntoIter = <SmallVec<[usize; N]> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<const N: usize> SortedSmallVec<N> {
+    fn new() -> Self {
+        Self(SmallVec::new())
+    }
+
+    const fn new_const() -> Self {
+        Self(SmallVec::new_const())
+    }
+
+    pub(crate) fn ones(&self) -> impl Iterator<Item = usize> + '_ {
+        self.0.iter().copied()
+    }
+
+    /// Insert the value if it's not already present in the vector.
+    /// Maintains a sorted order.
+    fn insert(&mut self, index: usize) {
+        match self.0.binary_search(&index) {
+            // element already present in the vector
+            Ok(_) => {}
+            Err(pos) => {
+                self.0.insert(pos, index);
+            }
+        }
+    }
+
+    /// Returns true if the vector contains the value.
+    fn contains(&self, index: usize) -> bool {
+        self.0.binary_search(&index).is_ok()
+    }
+
+    /// Returns true if the vector is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Empties the contents of the vector
+    pub(crate) fn clear(&mut self) {
+        self.0.clear()
+    }
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Adds all the elements from `other` into this vector. (skipping duplicates)
+    fn union_with(&mut self, other: &Self) {
+        let mut i = 0;
+        let mut j = 0;
+        while i < self.len() && j < other.len() {
+            if self.0[i] < other.0[j] {
+                i += 1;
+            } else if self.0[i] > other.0[j] {
+                self.0.insert(i, other.0[j]);
+                j += 1;
+            } else {
+                i += 1;
+                j += 1;
+            }
+        }
+        while j < other.len() {
+            self.0.push(other.0[j]);
+            j += 1;
+        }
+    }
```

Comment:
```suggestion
   fn union_with(&mut self, other: &Self) {
        let mut left = 0;
        for e in other {
            if left == self.len() {
                self.0.push(e);
                continue;
            }
            match self.0[left..].binary_search(e) {
                Ok(i) => {
                    left += i;
                    continue;
                }
                Err(i) => {
                    left += i;
                    self.0.insert(left, e)
                }
            }
        }
    }
```
I am curious, would something like this be faster?

Reply:
Yes that seems like it would work, worth trying!
In general the vectors should be pretty small though; queries/systems usually only access a handful of components.

Reply:
This would be O(n log(m)) where `n` is the length of the `self` and `m` the length of `other`. I don't think this would be really any faster (always benchmark though!), because `m` is usually small with small vectors.
I think the initial approach with O(m + n) is better, but performance can be approved by initiating a temporary smallvec with a capacity of m + n and filling that in appropriately. Then you can memswap this smallvec with self and this should reduce memory allocation.

Reply:
> This would be O(n log(m)) where `n` is the length of the `self` and `m` the length of `other`. I don't think this would be really any faster (always benchmark though!), because `m` is usually small with small vectors. I think the initial approach with O(m + n) is better
Unless I misunderstand something, why would this be O(n log(m))? I'd think this is O(log(n) + m), it iterates through every element of `other`, while binary searching `self` for each, until there are no more elements in `other`. If we reach the end of `self` first, we push the rest.
I would think this would be about equivalent if not better than the initial approach for small vectors, and better the larger they get.
That being said, I agree with always benchmarking though!
(Technically, some more comparisons could be saved by pulling out the contents of `binary_search` `insert` and the slice indexing, but I do not think that is worth the readability sacrifice, unless this is sensitive for perf.)